### DISCUSSION TITLE
Spine 4.2

### DISCUSSION THREAD

**Discussion by PCGEngineer** [axmolengine/axmol#1972] [General]
Just curious if there is an ETA for updating the axmol spine runtimes for 4.2. They added a lot of physics models since 4.1. For now just exporting 4.1 version from spine so no worries yet

**Comment by rh101** [comment]
Spine 4.2 is not much different from 4.1, so it was straightforward to add the support.

If you are able to test it out, then please get it from here: https://github.com/axmolengine/spine-axmol/tree/spine42

**Reply by paulocoutinhox** [reply]
Can you update readme to version 4.2, thanks.

**Reply by rh101** [reply]
> Can you update readme to version 4.2, thanks.

I'm waiting to verify that it works without issues first before changing that, since I haven't actually had the time to actually test it yet.

**Reply by rh101** [reply]
Just did a quick test for the 4.2 runtimes with the spine example projects, and it seems to be working.

**Reply by paulocoutinhox** [reply]
I replaced folder from axmol/extensions/spine to the 4.2 version and get these errors:

```
Undefined symbol: spine::InheritTimeline::setFrame(int, float, spine::Inherit)
Undefined symbol: spine::InheritTimeline::InheritTimeline(unsigned long, int)
Undefined symbol: spine::PhysicsConstraint::setToSetupPose()
Undefined symbol: spine::PhysicsConstraint::getBone()
Undefined symbol: spine::PhysicsConstraint::getData()
Undefined symbol: spine::PhysicsConstraint::PhysicsConstraint(spine::PhysicsConstraintData&, spine::Skeleton&)
Undefined symbol: spine::PhysicsConstraintData::PhysicsConstraintData(spine::String const&)
Undefined symbol: spine::PhysicsConstraintTimeline::PhysicsConstraintTimeline(unsigned long, unsigned long, int, spine::Property)
Undefined symbol: vtable for spine::PhysicsConstraintMixTimeline
Undefined symbol: vtable for spine::PhysicsConstraintMassTimeline
Undefined symbol: vtable for spine::PhysicsConstraintWindTimeline
Undefined symbol: vtable for spine::PhysicsConstraintResetTimeline
Undefined symbol: vtable for spine::PhysicsConstraintDampingTimeline
Undefined symbol: vtable for spine::PhysicsConstraintGravityTimeline
Undefined symbol: vtable for spine::PhysicsConstraintInertiaTimeline
Undefined symbol: vtable for spine::PhysicsConstraintStrengthTimeline
```

Do you set anything special?

Also, will need change here in future:
https://github.com/axmolengine/axmol/blob/4abc7f4a9453d418f0f89d7c7298e9214e8fb36a/extensions/spine/README.md?plain=1#L1

**Reply by rh101** [reply]
> I replaced folder from axmol/extensions/spine to the 4.2 version and get these errors:
> 

The reason it may not be working is because of `GLOB` usage in the cmake scripts, so your project is not picking up the new source files in the spine runtime.  The quickest way to check is to delete any existing build folders, and re-create them with the `axmol build ...` command etc..

Alternatively, and something which I personally think is the better option is that you do *not* replace the spine folder in Axmol, but instead disable the Spine version included in Axmol, and add it manually to your own project with something like this in your project `CMakeLists.txt`:

```
set(AX_ENABLE_EXT_SPINE OFF CACHE BOOL "Build extension spine" FORCE)

include(FetchContent)
FetchContent_Declare(
    spine
    GIT_REPOSITORY https://github.com/axmolengine/spine-axmol.git
    GIT_TAG spine42
    GIT_SHALLOW ON
    SOURCE_DIR "${CMAKE_CURRENT_LIST_DIR}/spine"
)
FetchContent_MakeAvailable(spine)
```

This way you would not be changing the contents of the Axmol folder, and make it easier for you to update to newer releases of Axmol in the future.

**Reply by paulocoutinhox** [reply]
If i do this, i get errors in others place, because folder is different:
<img width="1512" alt="image" src="https://github.com/axmolengine/axmol/assets/395096/b91b0a2f-a7c9-4f7f-b168-275ae3f1b8bd">

And i already add:

```
set(GAME_INC_DIRS
  "${CMAKE_CURRENT_SOURCE_DIR}/Source"
  "${CMAKE_CURRENT_SOURCE_DIR}/spine/runtime/include"
)
```

**Reply by paulocoutinhox** [reply]
I will try now with global include dir:
```
include_directories("${CMAKE_CURRENT_SOURCE_DIR}/spine/runtime/include")
```

**Reply by rh101** [reply]
You're almost right, but not quite there.  It is `cocostudio` that requires the paths, so it should be this:

```
target_include_directories(cocostudio
    PUBLIC ${CMAKE_CURRENT_LIST_DIR}/spine/runtime/include
    PUBLIC ${CMAKE_CURRENT_LIST_DIR}/spine/src
    PUBLIC ${_AX_ROOT}/extensions/Particle3D/src
)
```

If you checked `<axmol>/extensions/CMakeLists.txt`, you would have seen this:
```
if(AX_ENABLE_EXT_COCOSTUDIO)
  add_subdirectory(cocostudio)
  if(AX_ENABLE_EXT_SPINE)
    target_compile_definitions(cocostudio PRIVATE AX_ENABLE_EXT_SPINE=1)
    if(BUILD_SHARED_LIBS)
      target_link_libraries(cocostudio spine particle3d)
    else()
      target_include_directories(cocostudio
        PUBLIC ${CMAKE_CURRENT_LIST_DIR}/spine/runtime/include
        PUBLIC ${CMAKE_CURRENT_LIST_DIR}/spine/src
        PUBLIC ${CMAKE_CURRENT_LIST_DIR}/Particle3D/src
      )
    endif()
  endif()
```

So, if cocostudio is enabled, and Spine is enabled, then you can see what paths are required for cocostudio.

**Reply by paulocoutinhox** [reply]
Sure, relax, im only making something fast in a sample project to help you test, i don't read all the code, im only adding things to run and check spine lib with my animation. Im coding my game while help the tests here.

It is working for me:

![image](https://github.com/axmolengine/axmol/assets/395096/357992a7-1295-4805-a963-ff43606e8085)

On CMakeLists.txt after project name:

```
set(AX_ENABLE_EXT_SPINE OFF CACHE BOOL "Build extension spine" FORCE)
```

On CMakeLists.txt before code source:

```
include(FetchContent)

FetchContent_Declare(
    spine
    GIT_REPOSITORY https://github.com/axmolengine/spine-axmol.git
    GIT_TAG spine42
    GIT_SHALLOW ON
    SOURCE_DIR "${CMAKE_CURRENT_LIST_DIR}/spine"
)

FetchContent_MakeAvailable(spine)

if(AX_ENABLE_EXT_COCOSTUDIO AND NOT BUILD_SHARED_LIBS)
    target_include_directories(cocostudio
        PUBLIC ${CMAKE_CURRENT_LIST_DIR}/spine/runtime/include
        PUBLIC ${CMAKE_CURRENT_LIST_DIR}/spine/src
        PUBLIC ${_AX_ROOT}/extensions/Particle3D/src
    )
endif()
```

**Reply by paulocoutinhox** [reply]
Can you put 4.2 branch as default? https://github.com/axmolengine/spine-axmol/tree/spine42
and
Can you update here: https://github.com/axmolengine/axmol/blob/4abc7f4a9453d418f0f89d7c7298e9214e8fb36a/extensions/spine/README.md?plain=1#L1

Thanks.

**Reply by rh101** [reply]
> Can you put 4.2 branch as default? https://github.com/axmolengine/spine-axmol/tree/spine42

@halx99 Is it alright if we make spine42 the default branch in that repo?

**Reply by halx99** [reply]
> > Can you put 4.2 branch as default? https://github.com/axmolengine/spine-axmol/tree/spine42
> 
> @halx99 Is it alright if we make spine42 the default branch in that repo?

sure

**Comment by PCGEngineer** [comment]
Thanks for the fast updates. It looks like it may have broken 4.1 compatibility. It works for 4.2 so I'll update to that (no rush for me). Heres an example 4.1 json that it barfs on
[Foes.json](https://github.com/user-attachments/files/15945372/Foes.json).
Just throws a 'cannot load skeleton' assert. Seems to do that with all my 4.1 jsons. When I save it as 4.2 it works again.

**Reply by rh101** [reply]
Sorry, I don't understand what you mean?

The changes made to work with Axmol should not affect the loading of the skeleton files in any way.

Are you using the correct branch for the version of your skeleton files?  Branch spine41 for 4.1, and spine42 for the 4.2 skeleton files. I don't actually know if the 4.2 runtimes support loading skeleton files saved in versions older than 4.2, but that's a question you could ask the Esoteric developers.

**Reply by PCGEngineer** [reply]
Yep thats what I meant. I had some 4.1 files and 4.2 files mixed in on the same project. I updated all to 4.2 and re-exported from Spine and it worked. I'm suprised it can't handle past formats, but not impacting me now.  Thanks
